dev-envs/linux/vscode.sh

```diff
@@ -2,47 +2,122 @@
-rm "vscode-server-linux-${arch}.tar.gz"
+# VS Code
+vscode_metadata_url="https://update.code.visualstudio.com/api/versions/${VSCODE_VERSION}/linux-${arch}/stable"
+vscode_commit=$(curl -s "${vscode_metadata_url}" | jq -r .version)
+vscode_server_url="https://update.code.visualstudio.com/commit:${vscode_commit}/server-linux-${arch}/stable"
+vscode_cli_url="https://update.code.visualstudio.com/${VSCODE_VERSION}/cli-linux-${arch}/stable"
+
+# Cursor
+cursor_metadata_url="https://raw.githubusercontent.com/oslook/cursor-ai-downloads/refs/heads/main/version-history.json"
+# {"versions": [{"version": "...", "platforms": {...}}, ...]}
+cursor_artifact_url=$(curl -s "${cursor_metadata_url}" | jq -r --arg version "${CURSOR_VERSION}" '.versions[] | select(.version == $version) | .platforms["linux-x64"]')
+# https://downloads.cursor.com/production/<COMMIT>/<PLATFORM>/<ARCH>/<ARTIFACT>
+cursor_commit=$(echo "${cursor_artifact_url}" | cut -d'/' -f5)
+# The direct links always have `0` as the last character for some reason, perhaps an attempt at obfuscation
+cursor_commit_id="${cursor_commit%?}0"
+# Path from the official AppImage artifacts under `/usr/share/cursor/resources/app/product.json`
+cursor_server_url="https://cursor.blob.core.windows.net/remote-releases/${CURSOR_VERSION}-${cursor_commit_id}/vscode-reh-linux-${arch}.tar.gz"
+# Path from intercepting `wget` calls
+cursor_cli_url="https://cursor.blob.core.windows.net/remote-releases/${cursor_commit_id}/cli-alpine-${arch}.tar.gz"
+
+# Create shared extensions directory
+extensions_root_dir="${HOME}/.vscode-extensions"
+mkdir -p "${extensions_root_dir}"
+
+# Create function to install VS Code server
+function install_vscode() {
+    commit="$1"
+    root_dir_name="$2"
+    server_download_url="$3"
+    server_digest="$4"
+    cli_download_url="$5"
+    cli_digest="$6"
+
+    root_dir="${HOME}/${root_dir_name}"
+    install_dir="${root_dir}/cli/servers/Stable-${commit}"
+    unpack_dir="${install_dir}/server"
+
+    curl "${server_download_url}" -Lo "vscode_server.tar.gz"
+    digest=$(openssl dgst -sha256 "vscode_server.tar.gz" | cut -d' ' -f2)
+    if [[ "${digest}" != "${server_digest}" ]]; then
+        echo "Digest mismatch"
+        echo "Expected: ${server_digest}"
+        echo "Got: ${digest}"
+        exit 1
+    fi
+
+    mkdir -p "${unpack_dir}"
+    tar --no-same-owner -xf "vscode_server.tar.gz" -C "${unpack_dir}" --strip-components 1
+    rm "vscode_server.tar.gz"
+
+    # Inside the bin directory there will be a binary called `code-server`, `cursor-server`, etc.
+    # We need to capture the first word of the name (e.g. `code` or `cursor`) for use later
+    bin_dir="${unpack_dir}/bin"
+    binary_name=""
+    pushd "${bin_dir}"
+    for binary in *; do
+        if [[ "${binary}" =~ ^([^-]+)-server$ ]]; then
+            binary_name="${BASH_REMATCH[1]}"
+        fi
+    done
+    popd
```

Comment:
nit: I wasn't familiar with `pushd` and `popd`, might be clearer to just use `cd $OLDPWD` ?

```suggestion
    cd "${bin_dir}"
    for binary in *; do
        if [[ "${binary}" =~ ^([^-]+)-server$ ]]; then
            binary_name="${BASH_REMATCH[1]}"
        fi
    done
    cd "${OLDPWD}"
```